### PR TITLE
Expose admin externalView and idealState to client

### PR DESCRIPTION
**Pull request by david-hong**
Would be useful for clients to get the externalView and idealState of a (cluster, resource). Tested on some staging resources and seems to work fine.

### CODE REVIEW

File: admin.go [hunk 1]
```diff
@@ -853,3 +871,23 @@ func (adm Admin) isClusterSetup(cluster string) (bool, error) {
 		keyBuilder.stateModelDefs(),
 	)
 }
+
+// createDataAccessorForPath checks that the cluster and path is valid to create a DataAcessor
+func (adm Admin) createDataAccessorForPath(
```
Comment (zhixinwen) [inline admin.go]: I think it is fine to remove this func and do the check in an adhoc way. Other places in the repo follow the same way. If we want to refactor it, we should do it all. Maybe we can do it in another diff, if it is necessary.

File: admin.go [hunk 2]
```diff
@@ -785,29 +785,47 @@ func (adm Admin) ListInstances(cluster string) (string, error) {
...
-	accessor := newDataAccessor(adm.zkClient, builder)
 	r, err := accessor.Msg(instanceCfg)
 	if err != nil {
 		return "", err
 	}
 	return r.String(), nil
 }
 
+// ListIdealState shows a list of ideal states for the cluster resource
+func (adm Admin) ListIdealState(cluster string, resource string) (*model.IdealState, error) {
```
Comment (zhixinwen) [inline admin.go]: You can add it, but looks like you may not need to use it

File: admin_test.go [hunk 3]
```diff
@@ -305,6 +305,73 @@ func (s *AdminTestSuite) TestEnableDisableResource() {
...
+		t.Error("fail addResource")
+	}
+	// should fail, doesn't create externalView by default
+	if _, err := s.Admin.ListExternalView(cluster, resource); err != ErrNodeNotExist {
+		t.Error("must setup resource externalView before ListExternalView")
+	}
+
+	// create externalview
+	externalView := fmt.Sprintf("/%s/EXTERNALVIEW/%s", cluster, resource)
+	data := "{}" // empty json data
```
Comment (zhixinwen) [inline admin_test.go]: we should create some test data instead of an empty json.

File: admin_test.go [hunk 4]
```diff
@@ -305,6 +305,73 @@ func (s *AdminTestSuite) TestEnableDisableResource() {
...
+	s.Admin.zkClient.CreateDataWithPath(externalView, []byte(data))
+
+	if _, err := s.Admin.ListExternalView(cluster, resource); err != nil {
+		t.Error("expect OK")
```
Comment (zhixinwen) [inline admin_test.go]: check if you can read from model correctly. It is possible to have json parsing corrupted.

File: admin_test.go [hunk 5]
```diff
@@ -305,6 +305,73 @@ func (s *AdminTestSuite) TestEnableDisableResource() {
...
+	// fail when resource doesn't exist
+	if _, err := s.Admin.ListIdealState(cluster, resource); err != ErrNodeNotExist {
+		t.Error("must setup resource before ListIdealState")
+	}
+	// expect pass
+	if err := s.Admin.AddResource(cluster, "resource", 32, "MasterSlave"); err != nil {
+		t.Error("fail addResource")
+	}
+
+	if _, err := s.Admin.ListIdealState(cluster, resource); err != nil {
```
Comment (zhixinwen) [inline admin_test.go]: same as above

File: model/external_view.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+package model
+
+// ExternalView represents a Helix external view
+type ExternalView struct {
+	ZNRecord
+}
+
+// GetNumPartitions returns the number of partitions for the externalview
+func (s *ExternalView) GetNumPartitions() int {
```
Comment (zhixinwen) [inline model/external_view.go]: I think you need more than `GetNumPartitions`, right? Are you planning to add it later?

**Review by zhixinwen — COMMENTED** [review]
Generally looks good, a few comments to address. Also please try to test it in some staging env before landing.

File: model/external_view.go (hunk 6 above)
Comment (david-hong) [reply model/external_view.go]: Yeah, planning to add it later

**Review by zhixinwen — APPROVED** [review]
lgtm

**Pull request merged**